gcp/modules/gcp-stackdriver-monitoring/alert_k8s_snapshots_couchdb.tf

```diff
@@ -2,39 +2,47 @@ resource "google_monitoring_alert_policy" "k8s_snapshots_couchdb" {
   display_name = "Snapshots are being created for persistent volumes of CouchDB stateful set"
   combiner     = "OR"
 
-  conditions = [{
-    condition_absent {
-      filter   = "metric.type=\"logging.googleapis.com/user/k8s_snapshots.couchdb.snapshot_created\" resource.type=\"k8s_container\""
-      duration = "300s"
-
-      aggregations {
-        alignment_period   = "600s"
-        per_series_aligner = "ALIGN_SUM"
-        group_by_fields    = []
-      }
-    }
-
-    display_name = "Snapshot creation events are missing in K8s-snapshot logs for CouchDB"
-  },
+  conditions = [
     {
       condition_absent {
         filter   = "metric.type=\"logging.googleapis.com/user/compute.disks.createSnapshot\" resource.type=\"gce_disk\""
         duration = "300s"
 
         aggregations {
-          alignment_period   = "600s"
+          alignment_period   = "300s"
```

Comment:
Alignment window comes from backup deltas – we need to make sure that at least one snapshot per zone is created in 5 minutes. And duration setting makes alert policy more resilient to intermittent delays in snapshot processing. Are you suggesting that 10 minute window is too short or too long?

Reply:
See below for more details, I'm suggesting that duration should be longer than only one datapoint, otherwise it'll be prone to flapping. So with alignment of `1m` I would suggest duration of `6m+` (as the original data points are still expected to arrive in 5m intervals).

Reply:
There are more than a single snapshot creation event occurs per zone in 5 minutes, you can see the metric [rarely goes below 2](https://app.google.stackdriver.com/metrics-explorer?project=gpii-gcp-prd&timeSelection=%7B%22timeRange%22:%226h%22%7D&xyChart=%7B%22dataSets%22:%5B%7B%22timeSeriesFilter%22:%7B%22filter%22:%22metric.type%3D%5C%22logging.googleapis.com%2Fuser%2Fcompute.disks.createSnapshot%5C%22%20resource.type%3D%5C%22gce_disk%5C%22%22,%22perSeriesAligner%22:%22ALIGN_SUM%22,%22crossSeriesReducer%22:%22REDUCE_SUM%22,%22secondaryCrossSeriesReducer%22:%22REDUCE_NONE%22,%22minAlignmentPeriod%22:%22300s%22,%22groupByFields%22:%5B%22resource.label.%5C%22zone%5C%22%22%5D,%22unitOverride%22:%221%22%7D,%22targetAxis%22:%22Y1%22,%22plotType%22:%22LINE%22%7D%5D,%22options%22:%7B%22mode%22:%22COLOR%22%7D,%22constantLines%22:%5B%5D,%22timeshiftDuration%22:%220s%22,%22y1Axis%22:%7B%22label%22:%22y1Axis%22,%22scale%22:%22LINEAR%22%7D%7D&isAutoRefresh=true) for prd. I don't mind giving it more time to be on the safe side, but I observe no flapping whatsoever for the alert policy that is deployed to my environment (in 1+ week).

Reply:
> There are more than a single snapshot creation event occurs per zone in 5 minutes

Yeah, so that seems quite weird, and we probably need to improve the metric. I inspected it a bit more detail and there are several issues:

**a) 2 events seem to be created for every snapshot**, so our metric is effectively double of what we expect, check for example following filter in `prd`:
```
resource.type="gce_disk" AND protoPayload.methodName="v1.compute.disks.createSnapshot"
operation.id="operation-1575050294434-5987ff94d9b78-5f480ee8-de318034"
```

<img width="991" alt="Screenshot 2019-11-29 at 18 04 45" src="https://user-images.githubusercontent.com/3027227/69884957-c7a2f880-12d2-11ea-95c7-1f02e5c8a886.png">

It's 2 operations for one disk - one seems to be logged when the snapshot starts and other one when the snapshot is finished. There are some more differences in metadata, but interesting seems to be `operation.first: true` in first and the `operation.last: true` in the other one, perhaps that might help.

**b) the metric will currently  count errors when the snapshots fail too** - try following filter:
```
resource.type="gce_disk" AND protoPayload.methodName="v1.compute.disks.createSnapshot"
severity="ERROR"
```
it means that snapshot failed, but it will still be counted against the LBM which only has a filter `resource.type="gce_disk" AND protoPayload.methodName="v1.compute.disks.createSnapshot"`

**c) the metric will capture events from any snapshots** - although we don't have any other regular snapshots atm., if we ever do, the metric will be skewed. Perhaps it would make sense to filter further, otherwise, there's no guarantee that the snapshots even belong to CouchDB. Potentially interesting fields:
`protoPayload.authenticationInfo.principalEmail="gke-cluster-pod-k8s-snapshots@gpii-gcp-prd.iam.gserviceaccount.com"`, or `protoPayload.request.name="pv-database-storage-couchdb-couchdb-2-291119-181318"` although the later one is only present in the first (snapshot start) event.

> but I observe no flapping whatsoever

In addition to the above, have you tested situation like `k8s-snapshots` update, or GKE Node pool update when the "normal" operation of `k8s-snapshots` is disrupted for a whiile?

Reply:
@stepanstipl Thanks for the research, it is very helpful! GitHub makes it pretty difficult to keep inline conversations, I'll reply in thread instead.